cpp/evolution/v2/lib/binary/protocols.cc

```diff
@@ -0,0 +1,3957 @@
+    uint64_t uint_to_ulong;
+    yardl::binary::ReadInteger(stream_, uint_to_ulong);
+    value = static_cast<uint32_t>(uint_to_ulong);
+    break;
+  }
+  default:
+    yardl::binary::ReadInteger(stream_, value);
+    break;
+  }
+}
+
+void ProtocolWithChangesReader::ReadUintToFloatImpl(uint32_t& value) {
+  switch (version_) {
+  case Version::v1: {
+    float uint_to_float;
+    yardl::binary::ReadFloatingPoint(stream_, uint_to_float);
+    value = static_cast<uint32_t>(std::round(uint_to_float));
+    break;
+  }
+  default:
+    yardl::binary::ReadInteger(stream_, value);
+    break;
+  }
+}
+
+void ProtocolWithChangesReader::ReadUintToDoubleImpl(uint32_t& value) {
+  switch (version_) {
+  case Version::v1: {
+    double uint_to_double;
+    yardl::binary::ReadFloatingPoint(stream_, uint_to_double);
+    value = static_cast<uint32_t>(std::round(uint_to_double));
+    break;
+  }
+  default:
+    yardl::binary::ReadInteger(stream_, value);
+    break;
+  }
+}
+
+void ProtocolWithChangesReader::ReadFloatToDoubleImpl(float& value) {
+  switch (version_) {
+  case Version::v1: {
+    double float_to_double;
+    yardl::binary::ReadFloatingPoint(stream_, float_to_double);
+    value = static_cast<float>(float_to_double);
+    break;
+  }
+  default:
+    yardl::binary::ReadFloatingPoint(stream_, value);
+    break;
+  }
+}
+
+void ProtocolWithChangesReader::ReadIntToStringImpl(int32_t& value) {
+  switch (version_) {
+  case Version::v1: {
+    std::string int_to_string;
+    yardl::binary::ReadString(stream_, int_to_string);
+    value = std::stoi(int_to_string);
+    break;
+  }
+  default:
+    yardl::binary::ReadInteger(stream_, value);
+    break;
+  }
+}
+
+void ProtocolWithChangesReader::ReadUintToStringImpl(uint32_t& value) {
+  switch (version_) {
+  case Version::v1: {
+    std::string uint_to_string;
+    yardl::binary::ReadString(stream_, uint_to_string);
+    value = std::stoul(uint_to_string);
+    break;
+  }
+  default:
+    yardl::binary::ReadInteger(stream_, value);
+    break;
+  }
+}
+
+void ProtocolWithChangesReader::ReadLongToStringImpl(int64_t& value) {
+  switch (version_) {
+  case Version::v1: {
+    std::string long_to_string;
+    yardl::binary::ReadString(stream_, long_to_string);
+    value = std::stol(long_to_string);
+    break;
+  }
+  default:
+    yardl::binary::ReadInteger(stream_, value);
+    break;
+  }
+}
+
+void ProtocolWithChangesReader::ReadUlongToStringImpl(uint64_t& value) {
+  switch (version_) {
+  case Version::v1: {
+    std::string ulong_to_string;
+    yardl::binary::ReadString(stream_, ulong_to_string);
+    value = std::stoul(ulong_to_string);
+    break;
+  }
+  default:
+    yardl::binary::ReadInteger(stream_, value);
+    break;
+  }
+}
+
+void ProtocolWithChangesReader::ReadFloatToStringImpl(float& value) {
+  switch (version_) {
+  case Version::v1: {
+    std::string float_to_string;
+    yardl::binary::ReadString(stream_, float_to_string);
+    value = std::stof(float_to_string);
```

Comment:
Similarly, this doesn't give an informative error message if `float_to_string` is something like "Hello". It's hard for the user to know what the problem is, or even whether there is a bug in the tool.